k6/generic/change_config_maps.js

```diff
@@ -1,70 +1,467 @@
+// 2 requests per iteration (for change() func), so iteration rate is 1/2 of request rate
+const changeIPS = (__ENV.TARGET_RPS || 10) / 2
+// 1 request per iteration (for list() func)
+const listIPS = 1
 const kubeconfig = k8s.kubeconfig(__ENV.KUBECONFIG, __ENV.CONTEXT)
-const baseUrl = __ENV.BASE_URL
+
+// Metrics
+const diagnosticsBeforeGauge = new Gauge('diagnostics_before_churn');
+const diagnosticsDuringTrend = new Trend('diagnostics_during_churn');
+const diagnosticsAfterGauge = new Gauge('diagnostics_after_churn');
+const churnOpsCounter = new Counter('churn_operations_total');
+const churnOpsRate = new Trend('churn_operations_rate');
+let changeEvents = 0
 
 export const options = {
-    insecureSkipTLSVerify: true,
-    tlsAuth: [
-        {
-            cert: kubeconfig["cert"],
-            key: kubeconfig["key"],
-        },
-    ],
-
-    summaryTrendStats: ['avg', 'min', 'med', 'max', 'p(95)', 'p(99)', 'count'],
-
-    scenarios: {
-        change: {
-            executor: 'constant-arrival-rate',
-            exec: 'change',
-            preAllocatedVUs: vus,
-            duration: duration,
-            rate: rate,
-        },
+  insecureSkipTLSVerify: true,
+  tlsAuth: [
+    {
+      cert: kubeconfig["cert"],
+      key: kubeconfig["key"],
     },
-    thresholds: {
-        checks: ['rate>0.99']
-    }
+  ],
+
+  summaryTrendStats: ['avg', 'min', 'med', 'max', 'p(95)', 'p(99)', 'count'],
+
+  scenarios: {
+    preChurnDiagnostics: {
+      executor: 'shared-iterations',
+      exec: 'preChurnDiagnostics',
+      vus: 1,
+      iterations: 1,
+      startTime: '0s',
+      maxDuration: '5m',
+    },
+    change: {
+      executor: 'constant-arrival-rate',
+      exec: 'change',
+      preAllocatedVUs: vus,
+      duration: duration,
+      rate: changeIPS,
+      maxVUs: 20,
+      startTime: '5m'
+    },
+    list: {
+      executor: 'constant-arrival-rate',
+      exec: 'list',
+      preAllocatedVUs: vus,
+      duration: duration,
+      rate: listIPS,
+      maxVUs: 20,
+      startTime: '5m'
+    },
+    duringChurnDiagnostics: {
+      executor: 'constant-arrival-rate',
+      exec: 'duringChurnDiagnostics',
+      preAllocatedVUs: 1,
+      rate: 1, // Every N seconds
+      timeUnit: diagnosticsInterval,
+      duration: duration,
+      startTime: '5m',
+    },
+    postChurnDiagnostics: {
+      executor: 'shared-iterations',
+      exec: 'postChurnDiagnostics',
+      vus: 1,
+      iterations: 1,
+      startTime: `${parseInt(duration) + 5}m`,
+      maxDuration: '5m',
+    },
+  },
+  thresholds: {
+    http_req_failed: ['rate<=0.05'], // Allow 5% error rate during churn
+    http_req_duration: ['p(95)<=2000'], // 95% of requests should be below 2s
+    checks: ['rate>0.95'], // 95% success rate
+    churn_operations_total: [`count>=${changeIPS * parseInt(duration) * 0.9}`], // At least 90% of target ops
+    churn_operations_rate: ['med>=100']
```

Comment:
I do not understand this. Why 100, and why does `churn_operations_rate` / `churnOpsRate` never seem to be used? Am I missing anything?

Reply:
This was leftover from an attempt to track the # of "resource.change" operations/second but k6 does not provide a way to share state between VUs so this isnt feasible without some external datastore. Will remove.